File: sdks/java/src/main/java/io/intract/sdk/IntractContract.java
package io.intract.sdk;

import java.util.ArrayList;
import java.util.List;
import java.util.StringJoiner;

public class IntractContract {
    public String scope;
    public String intent;
    public int priority = 3;
    public String domain = "";
    public List<String> input = new ArrayList<>();
    public List<String> output = new ArrayList<>();
    public List<String> effect = new ArrayList<>();
    public List<String> forbid = new ArrayList<>();
    public List<String> require = new ArrayList<>();
    public List<String> validate = new ArrayList<>();
    public String meaning = "";

    public String inline(String prefix) {
        if (prefix == null || prefix.isBlank()) {
            prefix = "//";
        }

        List<String> parts = new ArrayList<>();
        parts.add("@intract.v1");
        parts.add("scope:" + scope);
        parts.add("intent:" + intent);
        parts.add("priority:" + priority);

        if (!domain.isBlank()) parts.add("domain:" + domain);
        if (!input.isEmpty()) parts.add("input:" + join(input));
        if (!output.isEmpty()) parts.add("output:" + join(output));
        if (!effect.isEmpty()) parts.add("effect:" + join(effect));
        if (!forbid.isEmpty()) parts.add("forbid:" + join(forbid));
        if (!require.isEmpty()) parts.add("require:" + join(require));
        if (!validate.isEmpty()) parts.add("validate:" + join(validate));
        if (!meaning.isBlank()) parts.add("meaning:\"" + meaning + "\"");

        return prefix + " " + String.join(" ", parts);
    }

    private String join(List<String> values) {
        StringJoiner joiner = new StringJoiner(",");
        for (String value : values) {
            joiner.add(value);
        }
        return joiner.toString();
    }
}

File: sdks/csharp/IntractContract.cs
namespace Intract.Sdk;

public sealed class IntractContract
{
    public string Scope { get; init; } = "function";
    public string Intent { get; init; } = "";
    public int Priority { get; init; } = 3;
    public string Domain { get; init; } = "";
    public string[] Input { get; init; } = [];
    public string[] Output { get; init; } = [];
    public string[] Effect { get; init; } = [];
    public string[] Forbid { get; init; } = [];
    public string[] Require { get; init; } = [];
    public string[] Validate { get; init; } = [];
    public string Meaning { get; init; } = "";

    public string Inline(string prefix = "//")
    {
        var parts = new List<string>
        {
            "@intract.v1",
            $"scope:{Scope}",
            $"intent:{Intent}",
            $"priority:{Priority}"
        };

        if (!string.IsNullOrWhiteSpace(Domain)) parts.Add($"domain:{Domain}");
        if (Input.Length > 0) parts.Add($"input:{string.Join(",", Input)}");
        if (Output.Length > 0) parts.Add($"output:{string.Join(",", Output)}");
        if (Effect.Length > 0) parts.Add($"effect:{string.Join(",", Effect)}");
        if (Forbid.Length > 0) parts.Add($"forbid:{string.Join(",", Forbid)}");
        if (Require.Length > 0) parts.Add($"require:{string.Join(",", Require)}");
        if (Validate.Length > 0) parts.Add($"validate:{string.Join(",", Validate)}");
        if (!string.IsNullOrWhiteSpace(Meaning)) parts.Add($"meaning:\"{Meaning}\"");

        return $"{prefix} {string.Join(" ", parts)}";
    }
}
